compiler/rustc_hir_typeck/src/lint.rs

```diff
@@ -0,0 +1,67 @@
+use rustc_hir::def_id::DefId;
+use rustc_hir::{HirId, HirIdSet};
+use rustc_middle::hir::place::{PlaceBase, PlaceWithHirId};
+use rustc_middle::mir::FakeReadCause;
+use rustc_middle::query::Providers;
+use rustc_middle::ty::{self, TyCtxt};
+use tracing::instrument;
+
+use crate::expr_use_visitor::{Delegate, ExprUseVisitor};
+
+#[derive(Default)]
+struct ExtractConsumingNodeDelegate {
+    nodes: HirIdSet,
+}
+
+impl<'tcx> Delegate<'tcx> for ExtractConsumingNodeDelegate {
+    #[instrument(level = "debug", skip(self))]
+    fn consume(&mut self, place_with_id: &PlaceWithHirId<'tcx>, _: HirId) {
+        match place_with_id.place.base {
+            PlaceBase::Rvalue => {
+                self.nodes.insert(place_with_id.hir_id);
+            }
+            PlaceBase::Local(id) => {
+                self.nodes.insert(id);
+            }
+            PlaceBase::Upvar(upvar) => {
+                self.nodes.insert(upvar.var_path.hir_id);
+            }
```

Comment:
Doesn't this count partial moves as full moves? Isn't that still not correct for drop ordering?

Reply:
Also, doesn't this not respect different paths correctly? What about consuming operations that happen on only one branch of a conditional path? I don't know if this sort of analysis is supported with `ExprUseVisitor`, which has no "flow" state.

Reply:
For now I don't have a good strategy towards that. How do we track partial moves and drops currently? Can we invoke machinery from borrow checker or somewhere else for such information?

I have switched to treat partial moves *not as full moves*. It could still get noisy but still safe.